bitcoin/src/witness_program.rs

```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
+            match WitnessProgram::from_str(INVALID_OP_1_PROGRAM_LENGTH) {
+                Ok(_) => assert_eq!(None, Some("Error, invalid program decoded successfully")),
+                Err(e) => {
+                    assert_eq!(WitnessProgramError::MismatchedProgramLength(38, 40), e)
+                }
+            }
+        }
+
+        #[test]
+        fn from_str_valid_op_1_program() {
+            match WitnessProgram::from_str(VALID_OP_1_PROGRAM) {
+                Ok(witness_program) => {
+                    assert_eq!(0x01, witness_program.version);
+                    assert_eq!(vec![0x75, 0x1e, 0x76, 0xe8, 0x19,
+                                    0x91, 0x96, 0xd4, 0x54, 0x94,
+                                    0x1c, 0x45, 0xd1, 0xb3, 0xa3,
+                                    0x23, 0xf1, 0x43, 0x3b, 0xd6,
+                                    0x75, 0x1e, 0x76, 0xe8, 0x19,
+                                    0x91, 0x96, 0xd4, 0x54, 0x94,
+                                    0x1c, 0x45, 0xd1, 0xb3, 0xa3,
+                                    0x23, 0xf1, 0x43, 0x3b, 0xd6
+                    ], witness_program.program);
+                }
+                Err(e) => assert_eq!(None, Some(e))
+            }
+        }
+
+        #[test]
+        fn from_str_invalid_hex_str() {
+            match WitnessProgram::from_str("001122zzxxyy") {
+                Ok(_) => assert_eq!(None, Some("Error, invalid hex string decoded successfully")),
+                Err(e) => assert_eq!(WitnessProgramError::ProgramDecodingError, e)
+            }
+        }
+    }
+
+    mod test_new {
+        use crate::witness_program::{WitnessProgram, WitnessProgramError};
+
+        const INVALID_VERSION_PROGRAM: &[u8] = &[0x19, 0x03, 0x00, 0x00, 0x00];
+
+        #[test]
+        fn new_invalid_version() {
+            match WitnessProgram::new(INVALID_VERSION_PROGRAM) {
+                Ok(_) => assert_eq!(None, Some("Error, program created with invalid version")),
+                Err(e) => assert_eq!(WitnessProgramError::InvalidVersion(0x19), e)
+            }
+        }
+
+        const INVALID_LENGTH_PROGRAM: &[u8] = &[0x19];
+
+        #[test]
+        fn new_invalid_length() {
+            match WitnessProgram::new(INVALID_LENGTH_PROGRAM) {
+                Ok(_) => assert_eq!(None, Some("Error, program created with invalid version")),
+                Err(e) => assert_eq!(WitnessProgramError::InvalidProgramLength(1), e)
+            }
+        }
+
+        const INVALID_LENGTH_FOR_VERSION: &[u8] = &[0x00, 0x0f, // Version 0, data length is incorrect
+            0x75, 0x1e, 0x76, 0xe8, 0x19,
+            0x91, 0x96, 0xd4, 0x54, 0x94,
+            0x1c, 0x45, 0xd1, 0xb3, 0xa3];
+
+        #[test]
+        fn new_invalid_program_length_for_version() {
+            match WitnessProgram::new(INVALID_LENGTH_FOR_VERSION) {
+                Ok(_) => assert_eq!(None, Some("Error, program created with invalid version")),
```

Comment:
This case should not be reachable

Reply:
See how to test for errors here:
https://github.com/ArgusHQ/wagyu/blob/v0.6.0/bitcoin/src/extended_private_key.rs#L845&L875